apis/Google.Devtools.AspNet/Google.Devtools.AspNet/CloudTrace/CloudTrace.cs

```diff
@@ -0,0 +1,138 @@
+using System.Threading.Tasks;
+using System.Web;
+
+namespace Google.Devtools.AspNet
+{
+    /// <summary>
+    ///  Google Trace for ASP.NET Applications.
+    /// </summary>
+    /// 
+    /// <example>
+    ///  public class Global : HttpApplication
+    ///  { 
+    ///     void Application_Start(object sender, EventArgs e)
+    ///     {
+    ///         CloudTrace.Initialize("some-project-id", this);
+    ///     }
+    ///  }
+    /// </example>
+    /// 
+    /// <example>
+    /// public void DoSomething()
+    /// {
+    ///   CloudTrace.GetCurrentTracer().StartSpan("DoSomething");
+    ///   ...
+    ///   CloudTrace.GetCurrentTracer().EndSpan();
+    /// }
+    /// </example>
+    /// 
+    /// <remarks>
+    /// Handles tracing for ASP.NET applications and sends data to the Stackdriver Trace APIReporting.
+    /// Docs: https://cloud.google.com/trace/docs/
+    /// </remarks>
+    public sealed class CloudTrace
+    {
+        private readonly string _projectId;
+        private readonly TraceIdFactory _traceIdfactory;
+        private readonly BufferingTraceConsumer _bufferingConsumer;
+        private readonly RateLimitingTraceOptionsFactory _rateFactory;
+        private readonly TraceHeaderTraceOptionsFactory _headerFactory;
+
+        private CloudTrace(string projectId, HttpApplication application, TraceConfiguration config = null,  Task<TraceServiceClient> client = null)
+        {
+            _projectId = GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+            GaxPreconditions.CheckNotNull(application, nameof(application));
+
+            // Create the default values if not set.
+            client = client ?? TraceServiceClient.CreateAsync();
+            config = config ?? TraceConfiguration.Create();
+
+            _traceIdfactory = TraceIdFactory.Create();
+            _bufferingConsumer = BufferingTraceConsumer.Create(GrpcTraceConsumer.Create(client));
+            _rateFactory = RateLimitingTraceOptionsFactory.Create(config);
+            _headerFactory = TraceHeaderTraceOptionsFactory.Create();
+
+            // Add event handlers to the application.
+            application.BeginRequest += new EventHandler(BeginRequest);
+            application.EndRequest += new EventHandler(EndRequest);
+        }
+
+        /// <summary>
+        /// Initialize tracing for this application.
+        /// </summary>
+        /// <param name="projectId">The Google Cloud Platform project ID.</param>
+        /// <param name="application">The Http application.</param>
+        /// <param name="config">Optional trace configuration, if unset the default will be used.</param>
+        /// <param name="client">Optional trace client, if unset the default will be used.</param>
+        public static void Initialize(string projectId, HttpApplication application, TraceConfiguration config = null,  Task<TraceServiceClient> client = null)
+        {
+            new CloudTrace(projectId, application, config, client);
```

Comment:
It seems very odd for the constructor to do the work here. Could we remove the application parameter from the constructor and change this method to:

```csharp
var trace = new CloudTrace(projectId, config, client);
application.BeginRequest += trace.BeginRequest;
application.EndRequest += trace.EndRequest;
```

That feels cleaner to me.